src/DistributeShifts.cpp

```diff
@@ -0,0 +1,200 @@
+#include "CSE.h"
+#include "CodeGen_Internal.h"
+#include "ConciseCasts.h"
+#include "FindIntrinsics.h"
+#include "IRMatch.h"
+#include "IRMutator.h"
+#include "Simplify.h"
+
+namespace Halide {
+namespace Internal {
+
+// Distribute constant RHS widening shift lefts as multiplies.
+// TODO: This is an extremely unfortunate mess. I think the better
+// solution is for the simplifier to distribute constant multiplications
```

Comment:
distributing constant multiplications goes in the wrong direction in terms of the simplifier's reduction order. It adds more ops! I think doing this transformation is very architecture and data-type specific, and we should do it as late as possible.

Reply:
This comment is copied over directly from HexagonOptimize!

Reply:
Please update it anyway. It's not a great comment, and you being the last person to touch it changes the meaning of "I"